hazelcast/src/main/java/com/hazelcast/map/impl/recordstore/DefaultRecordStore.java

```diff
@@ -1128,7 +1128,7 @@ public MapMergeResponse merge(MapMergeTypes<Object, Object> mergingEntry,
                 //  changed data and use that. Since this only matters for WAN-received merge events, we can avoid
                 //  additional overhead by checking provenance. Fixes HZ-3392, Backlog for merge changes: HZ-3397
                 boolean shouldMergeExpiration = provenance != CallerProvenance.WAN
-                        || valueComparator.isEqual(oldValue, mergingEntry.getValue(), serializationService);
+                        || valueComparator.isEqual(existingEntry.getRawValue(), mergingEntry.getRawValue(), serializationService);
```

Comment:
what If my map has OBJECT in memory format, will it still deserialize despite the raw value usage?

Reply:
If they're using the `OBJECT` in-memory format, then the `valueComparator` will be `ObjectValueComparator ` and that will invoke `toObject()` from the serialization service, yes.

However, this behaviour is expected for `OBJECT` as it is required for members to have classpath context in this case. The existing code already calls `valueComparator.isEqual` on the `oldValue` and `newValue` variables without issue, so these `toObject` calls should fall within expectations.

Please let me know if you think I've missed anything though 👍

Reply:
thx, i see the reason.

I have another question, we have `setTtl` in IMap api, how is it working in this case from the perspective of WAN? Because if you call `setTtl` then it is replicated over WAN, as far as i understand from the code, we don't reflect replicated ttl to the entry in target cluster? If this is broken thing a fix may be needed or otherwise if not documented we need to document the reasoning.

Reply:
TTL changes are replicated over WAN, that's what the original PR aimed to resolve (TTL was not replicated correctly when values were the same) - this additional value comparison is a "best effort" attempt to see if the `MergePolicy` applied actually changed the data or not.

As noted in my lengthy comment above the PR's code (ending line 1129), there's no way for us to see if a merge was rejected or not, only what the result is; currently the only way to "reject" a merge is to return the "old value" and use that instead of the merged value, so we're checking to see if it was "rejected" by comparing this old value, and if it has been rejected we don't want to WAN replicate the TTL metadata.

Reply:
ok i see setTtl is working with WAN, thx.